### PR TITLE
Kill stripe connect

### LINKED ISSUE #7872: Remove Stripe Connect
### What we should change and why (this is tech debt)

Now that we have moved to Stripe SCA we wish to remove Stripe Connect from the codebase to simplify and clean up our code.

All Stripe Connect code is be removed
No existing payment methods should reference this old payment method.
Any reports that reference this old payment method (Enterprise Fees report?) will now reference StripeSCA

### Context

All Stripe Support teams have handled the roll out of Stripe SCA so we can boldly make code changes and migrations here.

Another payment method was removed recently so this might be useful background:
#6231
and this resulting (and still unsolved) error is a mistake we should not repeat:
#7799

### Impact and timeline

While all users have been moved from Stripe Connect it is still possible for users to have chosen this payment method and be using it. Therefore we need to ensure we do a migration of any Stripe Connect payment methods to Stripe SCA.

### PR DESCRIPTION
**Pull request by apricot12**
#### What? Why?

Closes #7872 

#### Release notes
- Removes StripeConnect from code base and replaces it with StripeSCA wherever possible

Changelog Category: User facing changes | Technical changes

### CODE REVIEW

File: spec/jobs/subscription_confirm_job_spec.rb [hunk 1]
```diff
@@ -190,23 +190,6 @@
           job.send(:confirm_order!, order)
         end
       end
-
-      context "Stripe Connect" do
```
Comment (andrewpbrett) [inline spec/jobs/subscription_confirm_job_spec.rb:194]: We should actually keep these specs. Luis updated them to use the SCA payment method and just forgot to change the `context` line.

File: spec/controllers/spree/admin/orders/payments/payments_controller_spec.rb [hunk 2]
```diff
@@ -44,22 +44,6 @@
...
-            and_raise(Spree::Core::GatewayError.new("Payment Gateway Error"))
-        end
-
-        it "redirects to new payment page with flash error" do
-          spree_post :create, payment: params, order_id: order.number
-
-          redirects_to_new_payment_page_with_flash_error("Payment Gateway Error")
-          expect(order.reload.payments.last.state).to eq "checkout"
-        end
-      end
```
Comment (andrewpbrett) [inline spec/controllers/spree/admin/orders/payments/payments_controller_spec.rb]: Why is this redundant? What's the other spec that covers this same case?

File: spec/lib/stripe/profile_storer_spec.rb [hunk 3]
```diff
@@ -21,22 +21,9 @@ module Stripe
...
-        }
-
-        it "fetches the customer id and the card id from the correct response fields" do
-          profile_storer.create_customer_from_token
-
-          expect(payment.source.gateway_customer_profile_id).to eq customer_id
-          expect(payment.source.gateway_payment_profile_id).to eq card_id
-        end
-      end
-
```
Comment (andrewpbrett) [inline spec/lib/stripe/profile_storer_spec.rb]: I think we should leave this spec as well; it's not actually specific to Stripe Connect.

File: lib/stripe/profile_storer.rb [hunk 4]
```diff
@@ -67,11 +67,7 @@ def customer_profile_id(response)
     end
 
     def payment_profile_id(response)
-      if response.params['customer'] # Payment Intents API
-        response.params['id']
-      else
-        response.params['default_source'] || response.params['default_card']
-      end
```
Comment (andrewpbrett) [inline lib/stripe/profile_storer.rb]: Based on the comment, I would have expected that we should now return `response.params['id']` since our SCA gateway uses the Payment Intents API.

File: spec/support/request/stripe_stubs.rb [hunk 5]
```diff
@@ -28,11 +28,11 @@ def stub_payment_methods_post_request(request: { payment_method: "pm_123" }, res
   end
 
   # Attaches the payment method to the customer in the hub's stripe account
-  def stub_payment_method_attach_request
+  def stub_payment_method_attach_request(payment_method: "pm_123", customer: "cus_A123")
```
Comment (andrewpbrett) [inline spec/support/request/stripe_stubs.rb:31]: Nice 👍

**Review by andrewpbrett — COMMENTED** [review]
This looks good @apricot12 - there are some changes needed with the specs but it's getting close! I think this PR can replace #7978 rather than listing it as a dependency.

File: lib/stripe/profile_storer.rb (hunk 4 above)
Comment (andrewpbrett) [reply lib/stripe/profile_storer.rb]: Ah disregard this comment, I was going through each commit. Looks like you changed it back.

File: spec/jobs/subscription_confirm_job_spec.rb (hunk 1 above)
Comment (luisramos0) [reply spec/jobs/subscription_confirm_job_spec.rb:194]: correct :-)

can be changed to "Stripe SCA"

File: spec/controllers/spree/admin/orders/payments/payments_controller_spec.rb (hunk 2 above)
Comment (luisramos0) [reply spec/controllers/spree/admin/orders/payments/payments_controller_spec.rb]: yes, I think it should be kept.

File: spec/lib/stripe/profile_storer_spec.rb (hunk 3 above)
Comment (luisramos0) [reply spec/lib/stripe/profile_storer_spec.rb]: I agree. we can change the name to
context "when called from Stripe SCA" do

File: lib/stripe/profile_storer.rb [hunk 6]
```diff
@@ -65,7 +65,7 @@ def source_attrs_from(response)
     def customer_profile_id(response)
       response.params['customer'] || response.params['id']
     end
-
+
```
Comment (luisramos0) [inline lib/stripe/profile_storer.rb:68]: you can revert the changes in this file so the commit history is clean.
are these new spaces correct?

**Review by luisramos0 — APPROVED** [review]
LGTM!

**Comment by andrewpbrett** [comment]
@apricot12 what do you think of the suggestions here? I think it's pretty close to something we can merge.

**Comment by apricot12** [comment]
> @apricot12 what do you think of the suggestions here? I think it's pretty close to something we can merge. 

I think I can bring the tests back and push 👍

**Review by Matt-Yorkley — APPROVED** [review]
LGTM :+1:

**Comment by Matt-Yorkley** [comment]
@andrewpbrett and @luisramos0 with this migration do you think there any other bits of data that could be in the database related to payments or payment methods with `StripeConnect` that might be incompatible or might need updating? Like tokens or associated objects in a different format? Or not?

**Comment by luisramos0** [comment]
Hello everyone!
Matt, I cant think of anything.
Rachel, I think we need to test stripesca, at least the basics.

**Comment by apricot12** [comment]
> @apricot12 do you know what a tester should test here? We try some payments with Stripe connect then we stage the PR and look what happens to these specific orders?
> 
> Do we need to have a look at stripe SCA as well?

@RachL we need to see what happens to previous StripeConnect payments, as well as test stripeSCA. That should be enough.

**Comment by filipefurtad0** [comment]
Hey @apricot12 ,


**Before staging**
I was able to reproduce #7799 in staging-UK in which the error points out in this case to MIGS payment method, also recently removed:

https://openfoodnetwork.slack.com/archives/CEF14NU3V/p1630660254012100
https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/6131e69d45037800076b20b3?event_id=6131e69d007fa16670870000&i=sk&m=nw

Also before staging, I've created some orders with Stripe Connect, as to see how the app handles it after the removal of this payment gateway.

**After staging** we can see that:

1) the option to choose Stripe Connect (when creating/editing payment methods) has been removed

image

2) existing payment methods which used the Stripe Connect gateway have been migrated to use Stripe SCA gateway (below snippet from `/admin/payment_methods`)

image

3) orders previously created with Stripe Connect can still be accessed, namely to check the payment details:

image

4) These appear correctly under Order Cycle Management -> Payments Report displays the respective payment

image

5) Error 500: Regardless of this PR, running the Enterprise Fees Summary report still triggers the snail and related bugsnag error. This refers to MIGS so maybe we need a similar issue to #7799 and migration to address this as well.

6) Error 500: Running the Payment Reports -> Payment Totals broke the app:

https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/613207b231bb1f0007268b8b?event_id=613207b2007f99f415d40000&i=sk&m=nw

I'm not 100% sure this is related to the changes introduced here, but it seems likely. I'll double check on another server.

**Test cases not verified**

7) I've tried to place orders with Stripe Connect + card 42424... (valid card, no 3D auth. required). These payments succeed but the order appears with payment state "Authorization Required". This is not an issue, as we are removing Stripe Connect. However, it did prevent testing the **refund of Stripe Connect payments**, after this PR. I'm not sure this is a requirement, as this payment method is long deprecated, so hopefully we do not run into this scenario in production (i.e., refunding/voiding Stripe Connect Payments)

8) I think running the Enterprise Fees report should trigger the error #7799 on staging-AU as well. However, for the time being was not possible to verify this due to [this issue](https://github.com/openfoodfoundation/openfoodnetwork/issues/8133).

I would not push this to Ready to Go just yet. Perhaps best to clarify points 5-8. Thoughts?

File: db/migrate/20210817100257_rename_stripe_connect_payment_methods_to_stripe_sca.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+class RenameStripeConnectPaymentMethodsToStripeSca < ActiveRecord::Migration[6.1]
+  def change
+    Spree::PaymentMethod
+      .where(type: "Spree::Gateway::StripeConnect")
+      .update_all(type: "Spree::Gateway::StripeSCA")
```
Comment (mkllnk) [inline db/migrate/20210817100257_rename_stripe_connect_payment_methods_to_stripe_sca.rb:5]: Migrations last longer if they avoid using app code. In this case, it would be better to avoid `Spree::PaymentMethod` and execute simple SQL. While this statement looks simple, it doesn't actually replace all records:

```rb
puts Spree::PaymentMethod.where(type: "Spree::Gateway::StripeConnect").to_sql
SELECT "spree_payment_methods".* FROM "spree_payment_methods" WHERE \
"spree_payment_methods"."deleted_at" IS NULL AND "spree_payment_methods"."deleted_at" IS NULL AND \
"spree_payment_methods"."type" = 'Spree::Gateway::StripeConnect'
```

As you can see, there is a default scope that ignores deleted payment methods. If we changed the app code later and tried to load deleted payment methods then the app would crash on those old rows.

As a general rule: App code changes constantly but migrations should stay the same. So they should be independent and not be mixed.

```suggestion
    execute "UPDATE spree_payment_methods SET type = 'Spree::Gateway::StripeSCA' WHERE type = 'Spree::Gateway::StripeConnect'"
```

**Review by mkllnk — CHANGES_REQUESTED** [review]
Great work. I just found a catch in the migration that needs changing. I'm surprised that our good_migrations gem didn't pick that up. Did we deactivate it?

**Review by mkllnk — APPROVED** [review]

**Comment by filipefurtad0** [comment]
Hey @apricot12 ,

I'm not sure why but the migration seems to have worked for staging-FR, but not for staging-UK:


`SELECT * FROM spree_payment_methods WHERE type = 'Spree::Gateway::StripeConnect';`

```
 id  |             type              |                    name                     |                                                                                      description                                                                                       | active | environment |         created_at         |         updated_at         |         deleted_at         | display_on 
-----+-------------------------------+---------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+-------------+----------------------------+----------------------------+----------------------------+------------
  35 | Spree::Gateway::StripeConnect | Stripe payment to GP                        | Use the Stripe gateway to make your payment with a credit card.  A 3% fee will be added to your order.  (Note -  the Stripe account holder is 'Garden Party FLower Farm'.              | t      | production  | 2018-03-06 23:48:32.708637 | 2019-02-25 01:04:36.005906 | 2019-02-25 01:04:36.005906 | 
  59 | Spree::Gateway::StripeConnect | Credit Card Payment (via GP Stripe)         | Use the Stripe gateway to make your payment with a credit card.  A 3% fee will be added to your order.                                                                                 | t      | production  | 2019-02-25 01:08:05.640272 | 2020-02-10 20:28:12.816029 | 2020-02-10 20:28:12.816029 | 
  89 | Spree::Gateway::StripeConnect | Credit Card (on file) DEMO                  | Choose this option to leave our credit card with us.  Email or call us to confirm how much of a deposit you want to pay at this time,  and we'll remind you when the remainder is due. | t      | production  | 2019-12-27 16:53:53.671391 | 2020-01-12 14:27:55.708614 | 2020-01-12 14:27:55.708614 | 
 107 | Spree::Gateway::StripeConnect | Credit Card for members                     | We will waive the credit card processing fee for our members                                                                                                                           | t      | production  | 2020-03-27 20:22:21.753581 | 2020-04-06 18:37:05.649963 | 2020-04-06 18:37:05.649963 | 
 126 | Spree::Gateway::StripeConnect | second payment methods                      | Pay by credit card.                                                                                                                                                                    | t      | production  | 2020-04-10 22:13:36.903956 | 2020-09-23 01:16:37.995171 | 2020-09-23 01:16:37.995135 | 
 148 | Spree::Gateway::StripeConnect | Stripe                                      |                                                                                                                                                                                        | t      | production  | 2020-04-20 23:40:46.277114 | 2021-03-04 16:00:01.022845 | 2021-03-04 16:00:01.022801 | 
 424 | Spree::Gateway::StripeConnect | Debit / Credit Card through Stripe payments |                                                                                                                                                                                        | t      | production  | 2020-05-18 22:18:16.471388 | 2021-03-04 15:43:55.541807 | 2021-03-04 15:43:55.541767 | 
 481 | Spree::Gateway::StripeConnect | Quaker Springs Farm                         | e-transfer to quakerspringsfarm@gmail.com                                                                                                                                              | t      | production  | 2020-05-29 13:50:07.495973 | 2020-05-29 17:04:21.105734 | 2020-05-29 17:04:21.105734 | front_end
 556 | Spree::Gateway::StripeConnect | Damsma’s Honey                              |                                                                                                                                                                                        | t      | production  | 2020-06-07 11:47:00.835799 | 2020-06-07 11:51:31.236311 | 2020-06-07 11:51:31.236311 | 
 643 | Spree::Gateway::StripeConnect | Credit Card via Stripe                      | Note:  $3.00 for credit card processing will be added to your order at checkout                                                                                                        | t      | production  | 2020-07-01 14:25:35.861311 | 2020-12-29 18:46:23.102315 | 2020-12-29 18:46:23.102268 | 
(10 rows)
```

These are old, dummy payment methods. So I'm not sure this is caused by the state in which the DB was in staging-UK.

In both servers, the MIGS and Pin Payments payment methods still crash the Enterprise Fee Report, so I think this PR will not close #7799.

A smoke test of the Stripe SCA functionality revealed no issues... I did bump into something weird but appears rather concerning sessions (need to investigate further, and post in the appropriate issue).

So, everything seems OK in staging-FR :+1: For some reason, this is not the case in staging-UK. I'm not sure this will be an issue when deploying into production.

I'll wait for your feedback on this.

**Comment by filipefurtad0** [comment]
I think this did not work in staging-FR either, just before there was no StripeConnect payment method. I did the following test:
- created a Stripe Connect payment method
- logged into the server; ran the first query (below)
- staged the PR
- re-ran the query -> I was expecting no results - after the migration and server restart, which should happen during deploy, right? - but got the same result:

image


Also, `/admin/payment_methods/` and `/admin/enterprises/` pages became inaccessible.

**Comment by mkllnk** [comment]
@filipefurtad0 Do you know if the migration actually ran?

If the PR has been deployed in the past, the migration ran then and isn't re-applied unless renamed or the database is reset.

**Comment by filipefurtad0** [comment]
> If the PR has been deployed in the past, the migration ran then and isn't re-applied unless renamed or the database is reset.
Thanks this feedback @mkllnk, I was wondering about this too.

So, I'll run `delete from schema_migrations where version = '20210817100257'` in the psql shell and re-stage the PR. I'll try to find the corresponding version for staging-FR.

Thanks!

**Comment by filipefurtad0** [comment]
Alright! The migration worked: I've deleted the previous migration from both servers and can confirm the migration now worked (**on both servers**):

```
openfoodnetwork=> SELECT * FROM spree_payment_methods WHERE type = 'Spree::Gateway::StripeConnect';
 id | type | name | description | active | environment | created_at | updated_at | deleted_at | display_on 
----+------+------+-------------+--------+-------------+------------+------------+------------+------------
(0 rows)
```

I've noticed that for the previously created Stripe Connect payment methods the **account owner** is left empty:

- before migration

image

- after migration

image

I _think_ this does not cause any issues, as it removes such payment methods from the shopfront (if they would be still in use, which is unlikely), as these are not fully configured - account owner is mandatory. The user will experience an error on checkout, if - by coincidence - the payment method is displayed and the migration occurs during that time.

If a Stripe Connect payment method was associated in the past with a subscription, then the subscription won't run:
image

I'm not sure we have this case in production. What do you think?
Would it be possible to account for the **account owner** setting in this PR? (feedback label added)

Everything else looks fine: reports work, Stripe payments work :+1:

**Comment by apricot12** [comment]
> Alright! The migration worked: I've deleted the previous migration from both servers and can confirm the migration now worked (**on both servers**):
> 
> ```
> openfoodnetwork=> SELECT * FROM spree_payment_methods WHERE type = 'Spree::Gateway::StripeConnect';
>  id | type | name | description | active | environment | created_at | updated_at | deleted_at | display_on 
> ----+------+------+-------------+--------+-------------+------------+------------+------------+------------
> (0 rows)
> ```
> 
> I've noticed that for the previously created Stripe Connect payment methods the **Provider** is changed successfully but the **Account owner** is left empty:
> 
> * before migration
> 
> image
> 
> * after migration
> 
> image
> 
> I _think_ this does not cause any issues, as it removes such payment methods from the shopfront (if they would be still in use, which is unlikely), as these are not fully configured - account owner is mandatory. The user will experience an error on checkout, if - by coincidence - the payment method is displayed and the migration occurs during that time.
> 
> If a Stripe Connect payment method was associated in the past with a subscription, then the subscription won't run:
> image
> 
> I'm not sure we have this case in production. What do you think?
> Would it be possible to account for the **account owner** setting in this PR? (feedback label added)
> 
> Everything else looks fine: reports work, Stripe payments work 👍

@filipefurtad0 So after the migration is run, #7799 is solved? what about MIGS payment method? As for the account owner thing I'll have to take a look. But technically, it shouldn't be a problem.

**Comment by filipefurtad0** [comment]
Hey @apricot12 ,

Both in staging-UK and -FR the report is broken (error 500) when the page `/order_management/reports/enterprise_fee_summary/new` is accessed as superadmin; For hubs which have not used these payment methods the page renders as usual. I looked again at the DB in these servers and still found these entries:

image

The migration seems to have worked for `StripeConnect` but not for `Migs` and `Pin`, and so #7799 remains open.

**Comment by mkllnk** [comment]
Yes, we need a separate pull request with a migration for MIGS and PIN. I think that they need replacing with a simple cash method. Or can we just delete them from the database? That needs investigating.

**Comment by apricot12** [comment]
> Yes, we need a separate pull request with a migration for MIGS and PIN. I think that they need replacing with a simple cash method. Or can we just delete them from the database? That needs investigating.

I think that discussion already happened before and it was concluded that we want to Migrate them to something like cash?

**Comment by apricot12** [comment]
@filipefurtad0 I think this should be ready to go at this point? I dont think the payment method owner bit should cause any problems in prod. What do you think?

**Comment by filipefurtad0** [comment]
> I dont think the payment method owner bit should cause any problems in prod. What do you think?

Sure - that's a good idea. We can address the owner bit on a separate issue. Let's merge! :+1:
